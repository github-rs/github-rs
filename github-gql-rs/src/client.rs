// Tokio/Future Imports
use futures::future::ok;
use futures::{Future, Stream};
use tokio_core::reactor::Core;

use serde_json;
// Hyper Imports
use hyper::client::Client;
use hyper::StatusCode;
use hyper::{self, Headers};
#[cfg(feature = "rustls")]
use hyper_rustls::HttpsConnector;
#[cfg(feature = "rust-native-tls")]
use hyper_tls;
#[cfg(feature = "rust-native-tls")]
type HttpsConnector = hyper_tls::HttpsConnector<hyper::client::HttpConnector>;

// Serde Imports
use serde::de::DeserializeOwned;

// Lib Imports
use errors::*;
use mutation::Mutation;
use query::Query;
use IntoGithubRequest;

// Std Imports
use std::cell::RefCell;
use std::rc::Rc;

/// Struct used to make calls to the Github API.
pub struct Github {
    token: String,
    core: Rc<RefCell<Core>>,
    client: Rc<Client<HttpsConnector>>,
}

impl Clone for Github {
    fn clone(&self) -> Self {
        Self {
            token: self.token.clone(),
            core: self.core.clone(),
            client: self.client.clone(),
        }
    }
}

impl Github {
    /// Create a new Github client struct. It takes a type that can convert into
    /// a `String` (`&str` or `Vec<u8>` for example). As long as the function is
    /// given a valid API Token your requests will work.
    pub fn new<T>(token: T) -> Result<Self>
    where
        T: ToString,
    {
        let core = Core::new()?;
        let handle = core.handle();
        #[cfg(feature = "rustls")]
        let client = Client::configure()
            .connector(HttpsConnector::new(4, &handle))
            .build(&handle);
        #[cfg(feature = "rust-native-tls")]
        let client = Client::configure()
            .connector(HttpsConnector::new(4, &handle)?)
            .build(&handle);
        Ok(Self {
            token: token.to_string(),
            core: Rc::new(RefCell::new(core)),
            client: Rc::new(client),
        })
    }

    /// Get the currently set Authorization Token
    pub fn get_token(&self) -> &str {
        &self.token
    }

    /// Change the currently set Authorization Token using a type that can turn
    /// into an &str. Must be a valid API Token for requests to work.
    pub fn set_token<T>(&mut self, token: T)
    where
        T: ToString,
    {
        self.token = token.to_string();
    }

    /// Exposes the inner event loop for those who need
    /// access to it. The recommended way to safely access
    /// the core would be
    ///
    /// ```text
    /// let g = Github::new("API KEY");
    /// let core = g.get_core();
    /// // Handle the error here.
    /// let ref mut core_mut = *core.try_borrow_mut()?;
    /// // Do stuff with the core here. This prevents a runtime failure by
    /// // having two mutable borrows to the core at the same time.
    /// ```
    ///
    /// This is how other parts of the API are implemented to avoid causing your
    /// program to crash unexpectedly. While you could borrow without the
    /// `Result` being handled it's highly recommended you don't unless you know
    /// there is no other mutable reference to it.
    pub fn get_core(&self) -> &Rc<RefCell<Core>> {
        &self.core
    }

    pub fn query<T>(&mut self, query: &Query) -> Result<(Headers, StatusCode, Option<T>)>
    where
        T: DeserializeOwned,
    {
        self.run(query)
    }

    pub fn mutation<T>(&mut self, mutation: &Mutation) -> Result<(Headers, StatusCode, Option<T>)>
    where
        T: DeserializeOwned,
    {
        self.run(mutation)
    }

    fn run<T, I>(&mut self, request: &I) -> Result<(Headers, StatusCode, Option<T>)>
    where
        T: DeserializeOwned,
        I: IntoGithubRequest,
    {
        let mut core_ref = self.core.try_borrow_mut().chain_err(|| {
            "Unable to get mutable borrow \
             to the event loop"
        })?;
        let client = &self.client;
        let work = client
            .request(request.into_github_req(&self.token)?)
            .and_then(|res| {
                let header = res.headers().clone();
                let status = res.status();
                res.body()
                    .fold(Vec::new(), |mut v, chunk| {
                        v.extend(&chunk[..]);
                        ok::<_, hyper::Error>(v)
                    })
                    .map(move |chunks| {
                        if chunks.is_empty() {
                            Ok((header, status, None))
                        } else {
                            Ok((
                                header,
                                status,
                                Some(
                                    serde_json::from_slice(&chunks)
                                        .chain_err(|| "Failed to parse response body")?,
                                ),
                            ))
                        }
                    })
            });
        core_ref
            .run(work)
            .chain_err(|| "Failed to execute request")?
    }
}
