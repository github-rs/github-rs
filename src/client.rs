// Tokio/Future Imports
use futures::{ Future, Stream };
use futures::future::ok;
use tokio_core::reactor::Core;

// Hyper Imports
use hyper::{ Body, Uri, Method, Error };
use hyper::client::{ Client, Request };
use hyper::header::{ Authorization, Accept, ContentType, UserAgent, qitem };
use hyper::mime::Mime;
use hyper::status::StatusCode;
use hyper_tls::HttpsConnector;

// Serde Imports
use serde::Serialize;
use serde_json;

// Internal Library Imports
use users;
use misc;
use errors::*;
use Json;

/// Struct used to make calls to the Github API.
pub struct Github {
    token: String,
    core: Core,
}
new_type!(GetQueryBuilder);
new_type!(PutQueryBuilder);
new_type!(PostQueryBuilder);
new_type!(DeleteQueryBuilder);
new_type!(PatchQueryBuilder);
new_type!(Executor);


impl Github {
    pub fn new(token: &str) -> Self {
        Self {
            token: token.to_owned(),
            core: Core::new().unwrap(),
        }
    }
    pub fn get_token(&self) -> &str {
        &self.token
    }
    pub fn set_token(&mut self, token: &str) {
        self.token = token.to_owned();
    }
    pub fn get(&mut self) -> GetQueryBuilder {
        self.into()
    }
    pub fn put_empty(&mut self) -> PutQueryBuilder {
        self.into()
    }
    pub fn put<T>(&mut self, body: T) -> PutQueryBuilder
        where T: Serialize {
        let mut qb: PutQueryBuilder = self.into();
        match qb.request {
            Ok(mut qbr) => {
                let serialized = serde_json::to_vec(&body);
                match serialized {
                    Ok(json) => {
                        qbr.set_body(json);
                        qb.request = Ok(qbr);
                    },
                    Err(_) => {
                        qb.request = Err("Unable to serialize data to JSON".into());
                    }
                }
            },
            Err(_) => {},
        }
        qb
    }
    pub fn post<T>(&mut self, body: T) -> PostQueryBuilder
        where T: Serialize {
        let mut qb: PostQueryBuilder = self.into();
        match qb.request {
            Ok(mut qbr) => {
                let serialized = serde_json::to_vec(&body);
                match serialized {
                    Ok(json) => {
                        qbr.set_body(json);
                        qb.request = Ok(qbr);
                    },
                    Err(_) => {
                        qb.request = Err("Unable to serialize data to JSON".into());
                    }
                }
            },
            Err(_) => {},
        }

        qb
    }
    pub fn patch<T>(&mut self, body: T) -> PatchQueryBuilder
        where T: Serialize {
        let mut qb: PatchQueryBuilder = self.into();
        match qb.request {
            Ok(mut qbr) => {
                let serialized = serde_json::to_vec(&body);
                match serialized {
                    Ok(json) => {
                        qbr.set_body(json);
                        qb.request = Ok(qbr);
                    },
                    Err(_) => {
                        qb.request = Err("Unable to serialize data to JSON".into());
                    }
                }
            },
            Err(_) => {},
        }
        qb
    }
    pub fn delete<T>(&mut self, body: T) -> DeleteQueryBuilder
        where T: Serialize {
        let mut qb: DeleteQueryBuilder = self.into();
        match qb.request {
            Ok(mut qbr) => {
                let serialized = serde_json::to_vec(&body);
                match serialized {
                    Ok(json) => {
                        qbr.set_body(json);
                        qb.request = Ok(qbr);
                    },
                    Err(_) => {
                        qb.request = Err("Unable to serialize data to JSON".into());
                    }
                }
            },
            Err(_) => {},
        }
        qb
    }
    pub fn delete_empty(&mut self) -> DeleteQueryBuilder {
        self.into()
    }

}

impl<'a> GetQueryBuilder<'a> {
    func_client!(emojis, misc::get::Emojis<'a>);
    func_client!(rate_limit, misc::get::RateLimit<'a>);
    func_client!(user, users::get::User<'a>);
    func_client!(users, users::get::Users<'a>);
}

impl<'a> PutQueryBuilder<'a> {
    func_client!(user, users::put::User<'a>);
}

impl<'a> DeleteQueryBuilder<'a> {
    func_client!(user, users::delete::User<'a>);
}

impl<'a> PostQueryBuilder<'a> {
    func_client!(user, users::post::User<'a>);
}

impl<'a> PatchQueryBuilder<'a> {
    func_client!(user, users::patch::User<'a>);
}

impl<'a> Executor<'a> {

    pub fn execute(self) -> Result<(StatusCode, Option<Json>)> {
        let handle = self.core.handle();
        let work = Client::configure()
            .connector(HttpsConnector::new(4,&handle))
            .build(&handle)
            // All that type checking abuse culminates in this checking
            // if the request actually didn't fail being built!
            .request(self.request?)
            .and_then(|res| {
                let status = res.status().clone();
                res.body().fold(Vec::new(), |mut v, chunk| {
                    v.extend(&chunk[..]);
                    ok::<_, Error>(v)
                }).map(move |chunks| {
                    if chunks.is_empty() {
                        (status, None)
                    } else {
                        (status, Some(serde_json::from_slice(&chunks).unwrap()))
                    }
                })
            });

        self.core.run(work).chain_err(|| "Failed to execute request")
    }

}

// From derivations of Github to the given type using a certain
// request method
from!(GetQueryBuilder, Method::Get);
from!(PutQueryBuilder, Method::Put);
from!(PostQueryBuilder, Method::Post);
from!(PatchQueryBuilder, Method::Patch);
from!(DeleteQueryBuilder, Method::Delete);
