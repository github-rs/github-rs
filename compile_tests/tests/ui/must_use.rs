extern crate github_rs;

use github_rs::client::Github;

#[warn(unused_must_use)] // compiletest allows `unused` for UI tests, so we restore the warning here
fn main() {
    let g = Github::new("").unwrap();
    g.get().user().starred().owner("");
    g.get().gists();
}

