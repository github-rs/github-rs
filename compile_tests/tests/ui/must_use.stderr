warning: unused `github_rs::users::get::StarredOwner` which must be used: Incomplete query.
        This type by itself represents an incomplete query to the API.
        You need to build up a query that can be executed.
 --> $DIR/must_use.rs:8:5
  |
8 |     g.get().user().starred().owner("");
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: lint level defined here
 --> $DIR/must_use.rs:5:8
  |
5 | #[warn(unused_must_use)] // compiletest allows `unused` for UI tests, so we restore the warning here
  |        ^^^^^^^^^^^^^^^

warning: unused `github_rs::gists::get::Gists` which must be used: Unexecuted query.
        This type represents a query that can be executed. However,
        you have not chosen to execute the query for results at all.
        You must either execute the query or in some cases continue to
        build up the query.
 --> $DIR/must_use.rs:9:5
  |
9 |     g.get().gists();
  |     ^^^^^^^^^^^^^^^^

